File: packages/asdl-slots/src/asdl_slots/lifecycle/release_target.py
"""Shared target-release primitives for slot free and GC."""

from __future__ import annotations

import subprocess
from dataclasses import dataclass
from pathlib import Path
from typing import Literal

from asdl_core.git.git_gateway import GitGateway
from asdl_slots.inventory import SlotInventory, SlotRecord
from asdl_slots.lifecycle.outcomes import FreedSlot

ReleaseTargetFailureReason = Literal[
    "slot_not_assigned",
    "operation_in_progress",
    "dirty_worktree",
    "detach_failed",
]


@dataclass(frozen=True)
class ReleaseTargetFailure:
    reason: ReleaseTargetFailureReason
    slot_name: str
    branch_name: str
    worktree_path: Path
    operation: str | None = None
    detach_ref: str | None = None
    detach_error: str | None = None

    @property
    def error_type(self) -> str:
        if self.reason == "detach_failed":
            return "slot_allocation_error"
        return self.reason


def freed_slot_from_record(record: SlotRecord) -> FreedSlot:
    assert record.branch is not None
    return FreedSlot(
        slot_name=record.slot_name,
        branch_name=record.branch,
        worktree_path=record.path,
    )


def release_assigned_slot_target(
    git: GitGateway,
    inventory: SlotInventory,
    target: FreedSlot,
    trunk_branch: str,
) -> FreedSlot | ReleaseTargetFailure:
    """Recheck and detach one assigned target without owning caller policy."""
    record = inventory.find_by_slot(target.slot_name)
    if record is None or record.branch is None:
        return ReleaseTargetFailure(
            reason="slot_not_assigned",
            slot_name=target.slot_name,
            branch_name=target.branch_name,
            worktree_path=target.worktree_path,
        )

    if record.branch != target.branch_name:
        return ReleaseTargetFailure(
            reason="slot_not_assigned",
            slot_name=target.slot_name,
            branch_name=target.branch_name,
            worktree_path=record.path,
        )

    if record.operation is not None:
        return ReleaseTargetFailure(
            reason="operation_in_progress",
            slot_name=target.slot_name,
            branch_name=target.branch_name,
            worktree_path=record.path,
            operation=record.operation,
        )

    if git.has_uncommitted_changes(record.path):
        return ReleaseTargetFailure(
            reason="dirty_worktree",
            slot_name=target.slot_name,
            branch_name=target.branch_name,
            worktree_path=record.path,
        )

    try:
        git.detach_head(record.path, trunk_branch)
    except subprocess.CalledProcessError as exc:
        stderr = exc.stderr.strip() if exc.stderr else str(exc)
        return ReleaseTargetFailure(
            reason="detach_failed",
            slot_name=target.slot_name,
            branch_name=target.branch_name,
            worktree_path=record.path,
            detach_ref=trunk_branch,
            detach_error=stderr,
        )

    return freed_slot_from_record(record)

File: ts/packages/capabilities/slot/src/lifecycle/release-target.ts
import type { SlotRepositoryGateway } from "../gateways/repository.ts";
import { findBySlot, type SlotInventory, type SlotRecord } from "../inventory.ts";

export interface FreedSlot {
	slotName: string;
	branchName: string;
	worktreePath: string;
}

export type ReleaseTargetFailureReason =
	| "slot-not-assigned"
	| "operation-in-progress"
	| "dirty-worktree"
	| "detach-failed";

export interface ReleaseTargetFailure {
	reason: ReleaseTargetFailureReason;
	slotName: string;
	branchName: string;
	worktreePath: string;
	operation: string | null;
	detachRef: string | null;
	detachError: string | null;
	errorType: string;
}

export function freedSlotFromRecord(record: SlotRecord): FreedSlot {
	if (record.branch === null) throw new Error(`slot ${record.slotName} is not assigned`);
	return { slotName: record.slotName, branchName: record.branch, worktreePath: record.path };
}

export interface ReleaseAssignedSlotTargetOptions {
	git: SlotRepositoryGateway;
	inventory: SlotInventory;
	target: FreedSlot;
	trunkBranch: string;
}

export async function releaseAssignedSlotTarget(
	options: ReleaseAssignedSlotTargetOptions,
): Promise<FreedSlot | ReleaseTargetFailure> {
	const record = findBySlot(options.inventory, options.target.slotName);
	if (record === null || record.branch === null || record.branch !== options.target.branchName)
		return failure(
			"slot-not-assigned",
			options.target,
			record === null ? {} : { worktreePath: record.path },
		);
	if (record.operation !== null)
		return failure("operation-in-progress", options.target, {
			worktreePath: record.path,
			operation: record.operation,
		});
	if (await options.git.hasUncommittedChanges(record.path))
		return failure("dirty-worktree", options.target, { worktreePath: record.path });
	const detachFailure = await options.git.detachHead(record.path, options.trunkBranch);
	if (detachFailure !== null)
		return failure("detach-failed", options.target, {
			worktreePath: record.path,
			detachRef: options.trunkBranch,
			detachError: detachFailure.message,
		});
	return freedSlotFromRecord(record);
}

function failure(
	reason: ReleaseTargetFailureReason,
	target: FreedSlot,
	options: {
		worktreePath?: string;
		operation?: string | null;
		detachRef?: string;
		detachError?: string;
	},
): ReleaseTargetFailure {
	return {
		reason,
		slotName: target.slotName,
		branchName: target.branchName,
		worktreePath: options.worktreePath ?? target.worktreePath,
		operation: options.operation ?? null,
		detachRef: options.detachRef ?? null,
		detachError: options.detachError ?? null,
		errorType: reason === "detach-failed" ? "slot-allocation-error" : reason,
	};
}
